File: AC/library/ilib/web/ffi/web_ffi.java
import java.awt.Desktop;
import java.net.URI;

public class WebLib {
    public static void open(String link) {
        openURL("https://" + link);
    }

    public static void fileOpen(String file) {
        openURL("file://" + file);
    }

    public static void popen(String rawLink) {
        openURL(rawLink);
    }

    public static void ropen(String identifier) {
        openURL("https://" + identifier + ".com");
    }

    public static void browser() {
        openURL("https://www.google.com");
    }

    public static boolean pdf(String pdf) {
        if (pdf.endsWith(".pdf")) {
            fileOpen(pdf);
            return true;
        }
        return false;
    }

    public static boolean text(String text) {
        String[] validExts = {".txt", ".md", ".bashrc", ".zshrc"};
        for (String ext : validExts) {
            if (text.endsWith(ext)) {
                fileOpen(text);
                return true;
            }
        }
        return false;
    }

    public static boolean inspect(String program) {
        String[] validExts = {".py", ".java", ".js", ".c", ".cpp", ".v", ".ac", ".s", ".go", ".rs", ".sh", ".html", ".sql"};
        for (String ext : validExts) {
            if (program.endsWith(ext)) {
                fileOpen(program);
                return true;
            }
        }
        return false;
    }

    public static void acPage() {
        openURL("https://aclang.vercel.app");
    }

    public static String pageGet(String url) {
        // Stub - would need java.net.http for real implementation
        return null;
    }

    public static void help() {
        String helpText = "=== Web Library Functions ===\n" +
            "web.open(link)           - Opens https://{link}\n" +
            "web.fileOpen(file)       - Opens file://{file}\n" +
            "web.popen(rawLink)       - Opens {rawLink} as-is\n" +
            "web.ropen(search)        - Opens https://{search}.com\n" +
            "web.browser()            - Opens https://www.google.com\n" +
            "web.pdf(pdf)             - Opens file://{pdf}, must be .pdf\n" +
            "web.text(text)           - Opens file://{text}, must be .txt, .md, .bashrc, or .zshrc\n" +
            "web.inspect(program)     - Opens file://{program}, must be source code\n" +
            "web.acPage()             - Opens the AC language website\n" +
            "web.pageGet(link)        - Gets HTML contents of a webpage\n" +
            "web.help()               - Shows this help message";
        System.out.println(helpText);
    }

    private static void openURL(String url) {
        try {
            if (Desktop.isDesktopSupported()) {
                Desktop.getDesktop().browse(new URI(url));
            } else {
                openURLFallback(url);
            }
        } catch (Exception e) {
            openURLFallback(url);
        }
    }

    private static void openURLFallback(String url) {
        try {
            String osName = System.getProperty("os.name").toLowerCase();
            if (osName.contains("win")) {
                Runtime.getRuntime().exec("cmd /c start " + url);
            } else if (osName.contains("mac")) {
                Runtime.getRuntime().exec("open " + url);
            } else if (osName.contains("linux") || osName.contains("unix")) {
                Runtime.getRuntime().exec("xdg-open " + url);
            }
        } catch (Exception e) {
            System.err.println("Could not open URL: " + e.getMessage());
        }
    }
}

File: AC/library/ilib/web/ffi/web_ffi.go
package main

import (
	"fmt"
	"os/exec"
	"runtime"
	"strings"
)

var web = struct {
	open      func(string)
	fileOpen  func(string)
	popen     func(string)
	ropen     func(string)
	browser   func()
	pdf       func(string) bool
	text      func(string) bool
	inspect   func(string) bool
	acPage    func()
	pageGet   func(string) string
	help      func()
}{
	open: func(link string) {
		openURL("https://" + link)
	},
	fileOpen: func(file string) {
		openURL("file://" + file)
	},
	popen: func(rawLink string) {
		openURL(rawLink)
	},
	ropen: func(identifier string) {
		openURL("https://" + identifier + ".com")
	},
	browser: func() {
		openURL("https://www.google.com")
	},
	pdf: func(pdf string) bool {
		if strings.HasSuffix(pdf, ".pdf") {
			web.fileOpen(pdf)
			return true
		}
		return false
	},
	text: func(text string) bool {
		validExts := []string{".txt", ".md", ".bashrc", ".zshrc"}
		for _, ext := range validExts {
			if strings.HasSuffix(text, ext) {
				web.fileOpen(text)
				return true
			}
		}
		return false
	},
	inspect: func(program string) bool {
		validExts := []string{".py", ".java", ".js", ".c", ".cpp", ".v", ".ac", ".s", ".go", ".rs", ".sh", ".html", ".sql"}
		for _, ext := range validExts {
			if strings.HasSuffix(program, ext) {
				web.fileOpen(program)
				return true
			}
		}
		return false
	},
	acPage: func() {
		openURL("https://aclang.vercel.app")
	},
	pageGet: func(url string) string {
		// Stub - would need net/http for real implementation
		return ""
	},
	help: func() {
		helpText := `=== Web Library Functions ===
web.open(link)           - Opens https://{link}
web.file_open(file)      - Opens file://{file}
web.popen(raw_link)      - Opens {raw_link} as-is
web.ropen(search)        - Opens https://{search}.com
web.browser()            - Opens https://www.google.com
web.pdf(pdf)             - Opens file://{pdf}, must be .pdf
web.text(text)           - Opens file://{text}, must be .txt, .md, .bashrc, or .zshrc
web.inspect(program)     - Opens file://{program}, must be source code
web.ac_page()            - Opens the AC language website
web.page_get(link)       - Gets HTML contents of a webpage
web.help()               - Shows this help message`
		fmt.Println(helpText)
	},
}

func openURL(url string) {
	var cmd string
	var args []string

	switch runtime.GOOS {
	case "darwin":
		cmd = "open"
	case "windows":
		cmd = "cmd"
		args = []string{"/c", "start"}
	default:
		cmd = "xdg-open"
	}

	args = append(args, url)
	exec.Command(cmd, args...).Start()
}
